ffi/rtc.lua

```diff
@@ -0,0 +1,196 @@
+--[[--
+Module for interfacing with the RTC (real time clock).
+
+This module provides the ability to schedule wakeups through RTC.
+
+@module ffi.rtc
+]]
+
+local ffi = require("ffi")
+local bor = bit.bor
+local C = ffi.C
+
+local dummy = require("ffi/posix_h")
+local dummy = require("ffi/rtc_h")
+
+-----------------------------------------------------------------
+
+local RTC = {
+    _wakeup_scheduled = false,
+    _wakeup_scheduled_ptm = nil,
+}
+
+--[[
+Adds seconds to epoch.
+
+@int seconds_from_now
+@treturn cdata epoch
+--]]
+function RTC:secondsFromNowToEpoch(seconds_from_now)
+    local t = ffi.new("time_t[1]")
+    t[0] = C.time(nil)
+    t[0] = t[0] + seconds_from_now
+    local epoch = C.mktime(C.localtime(t))
+    return epoch
+end
+
+--[[
+Set wakeup alarm.
+
+@int epoch If you want to set to an amount of time from now, process your value with @{secondsFromNowToEpoch}
+@enabled bool Whether the call enables or disables the alarm. Defaults to true.
+
+@treturn bool Success.
+@treturn re Error code (if any).
+@treturn err Error string (if any).
+--]]
+function RTC:setWakeupAlarm(epoch, enabled)
+    enabled = (enabled ~= nil) and enabled or true
+
+    local ptm = C.gmtime(ffi.new("int[1]", epoch))
+    self._wakeup_scheduled_ptm = ptm
+
+    local wake = ffi.new("struct rtc_wkalrm")
+    wake.time.tm_sec = ptm.tm_sec
+    wake.time.tm_min = ptm.tm_min
+    wake.time.tm_hour = ptm.tm_hour
+    wake.time.tm_mday = ptm.tm_mday
+    wake.time.tm_mon = ptm.tm_mon
+    wake.time.tm_year = ptm.tm_year
+    -- wday, yday, and isdst fields are unused by Linux
+    wake.time.tm_wday = -1
+    wake.time.tm_yday = -1
+    wake.time.tm_isdst = -1
+
+    wake.enabled = enabled and 1 or 0
+
+    local err
+    local rtc0 = C.open("/dev/rtc0", bor(C.O_RDONLY, C.O_NONBLOCK))
+    if rtc0 == -1 then
+        err = ffi.string(C.strerror(ffi.errno()))
+        print("setWakeupAlarm open /dev/rtc0", rtc0, err)
+    end
+    local re = C.ioctl(rtc0, C.RTC_WKALM_SET, wake)
+    if re == -1 then
+        err = ffi.string(C.strerror(ffi.errno()))
+        print("setWakeupAlarm ioctl RTC_WKALM_SET", re, err)
+    end
+    re = C.close(rtc0)
+    if re == -1 then
+        err = ffi.string(C.strerror(ffi.errno()))
+        print("setWakeupAlarm close /dev/rtc0", re, err)
+    end
+
+    if re == 0 then
+        if enabled then
```

Comment:
you go on doing things above even if the first one failed? expecting all the others will fail so you still get `re==-1` in the end?
I would expect a `return nil, re, err` in each of the `ifs`

Reply:
Good point.

Reply:
Although technically the close doesn't mean it's not scheduled.